docs/topics/jvm-api-guidelines-backward-compatibility.md

```diff
@@ -0,0 +1,420 @@
+there is no way to be sure that your needs won't change in the future. So, all the issues with data classes only exist 
+when you decide to change such a class.
+
+Firstly, the considerations from the previous section [Don't add arguments to existing API functions](#don-t-add-arguments-to-existing-api-functions) 
+also apply to the constructor as it is also a method. Secondly, even if you add secondary constructors, it won't solve 
+the compatibility problem. Let's look at the following data class:
+
+```kotlin
+data class User(
+    val name: String,
+    val email: String
+)
+```
+
+For example, over some time, you understand that users should go through an activation procedure, so you want to add 
+a new field, `active` with a default value equal to `true`. This new field should allow the existing code to work 
+mostly without changes.
+
+As it was already discussed in the [section above](#don-t-add-arguments-to-existing-api-functions), you can't just add 
+a new field like this:
+
+```kotlin
+data class User(
+    val name: String,
+    val email: String,
+    val active: Boolean = true
+)
+```
+
+Because this change is **not binary compatible**.
+
+Let's add a new constructor that accepts only two arguments and calls the primary constructor with a third default argument:
+
+```kotlin
+data class User(
+    val name: String,
+    val email: String,
+    val active: Boolean = true
+) {
+    constructor(name: String, email: String): 
+            this(name, email, active = true)
+}
+```
+
+This time there are two constructors, and a signature of one of them matches the constructor of the class before the change:
+
+```java
+public User(java.lang.String, java.lang.String);
+```
+
+But the issue is not with the constructor, it's with the `copy` function. Its signature has changed from:
+
+```java
+public final User copy(java.lang.String, java.lang.String);
+```
+
+To:
+
+```java
+public final User copy(java.lang.String, java.lang.String, boolean);
+```
+
+And it has made the code **binary incompatible**.
+
+Of course, it's possible just to add a property inside the data class, but it removes all the bonuses of it being 
+a data class. Therefore, it's better not to use data classes in your API because almost any change in them breaks source, 
+binary, or behavioral compatibility.
+
+If you have to use a data class for whatever reason, you have to override the constructor and the `copy()` method. 
+In addition, if you add a field into the class's body, you have to override the `hashCode()` and `equals()` methods.
+
+> It's always an incompatible change to swap the order of arguments because of `componentX()` methods. It breaks source 
+> compatibility and probably will break binary compatibility too.
+>
+{type="warning"}
+
+### Don't make return types narrower
+
+Sometimes, especially when you don't use [explicit API mode](whatsnew14.md#explicit-api-mode-for-library-authors), 
```

Comment:
Or we can link to our chapter Explicit API mode, but there we don't tell what explicit API mode is.